src/libNode/Node.cpp

```diff
@@ -2408,6 +2427,199 @@ bool Node::ProcessDoRejoin(const bytes& message, unsigned int offset,
+                "[update shard guard] Node to be updated is current node. No "
+                "update needed.");
+    return false;
+  }
+
+  uint64_t currentDSEpochNumber =
+      m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() + 1;
+
+  if (dsEpochNumber != currentDSEpochNumber) {
+    LOG_GENERAL(
+        WARNING,
+        "Update of shard guard network info failure  - dsepoch in message: "
+            << dsEpochNumber
+            << " does not match current dsepoch: " << currentDSEpochNumber);
+    return false;
+  }
+
+  if (!Guard::GetInstance().IsNodeInShardGuardList(shardGuardPubkey)) {
+    LOG_GENERAL(WARNING,
+                "Update of shard guard requested by non-shard guard node, So "
+                "ignore the request");
+    return false;
+  }
+  // I am lookup node
+  if (LOOKUP_NODE_MODE) {
+    m_mediator.m_ds->UpdateShardNodeNetworkInfo(shardGuardNewNetworkInfo,
+                                                shardGuardPubkey);
+    if (!BlockStorage::GetBlockStorage().PutShardStructure(
+            m_mediator.m_ds->m_shards, 0)) {
+      LOG_GENERAL(WARNING, "BlockStorage::PutShardStructure failed");
+    }
+  }
+  // I am sharded node and requestor is also from my shard
+  else if (m_mediator.m_ds->m_mode == DirectoryService::IDLE) {
+    if (!IsShardNode(shardGuardPubkey)) {
+      LOG_GENERAL(WARNING, "PubKey of sender "
+                               << from
+                               << " does not match any of my shard members");
+      return false;
+    }
+    // update requestor's ( ShardGuard ) new IP
+    lock_guard<mutex> g(m_mutexShardMember);
+
+    unsigned int indexOfShardNode;
+    for (indexOfShardNode = 0;
+         indexOfShardNode < Guard::GetInstance().GetNumOfShardGuard();
+         indexOfShardNode++) {
+      if (m_myShardMembers->at(indexOfShardNode).first == shardGuardPubkey) {
+        LOG_GENERAL(
+            INFO, "[update shard guard] shard guard to be updated is at index "
+                      << indexOfShardNode << " "
+                      << m_myShardMembers->at(indexOfShardNode).second << " -> "
+                      << shardGuardNewNetworkInfo);
+        m_myShardMembers->at(indexOfShardNode).second =
+            shardGuardNewNetworkInfo;
+        if (BROADCAST_GOSSIP_MODE) {
+          // Update peer info for gossip
+          P2PComm::GetInstance().UpdatePeerInfoInRumorManager(
+              shardGuardNewNetworkInfo, shardGuardPubkey);
+        }
+
+        // Put the sharding structure to disk
+        if (!BlockStorage::GetBlockStorage().PutShardStructure(
+                m_mediator.m_ds->m_shards, m_mediator.m_node->m_myshardId)) {
+          LOG_GENERAL(WARNING, "BlockStorage::PutShardStructure failed");
+        }
+        break;
+      }
+    }
+  }
+  // I am ds node and requestor is from one of shards
+  else if (m_mediator.m_ds->m_mode != DirectoryService::IDLE) {
+    if (!m_mediator.m_ds->CheckIfShardNode(shardGuardPubkey)) {
```

Comment:
Here we make an initial search through `m_shards` and then in line 2610 we also go through `m_shards` to make the actual replacement. Maybe we can skip this initial search and just try to find and replace immediately? If it's not found during replacement, you can log the "does not match" message as well.